### PR TITLE
♻️ [#200] Configure Autorisaties API connection via setup-configuration

### LINKED ISSUE #200: Add support for Autorisaties-API via django-setup-configuration
- [x] reuse service configurationstep in ON
- [x] add step to configure AuthorizationsConfig
- [x] Add configurationstep to commonground api common for JWTSecret
- [x] Reuse that configurationstep in ON

### PR DESCRIPTION
**Pull request by stevenbal**
Closes #200

### CODE REVIEW

File: src/nrc/tests/commands/test_setup_configuration.py [hunk 1]
```diff
@@ -1,36 +1,25 @@
...
-    AUTHORIZATION_CONFIG_ENABLE=True,
-    AUTORISATIES_API_ROOT="https://oz.example.com/autorisaties/api/v1/",
-    NOTIF_OPENZAAK_CLIENT_ID="notif-client-id",
-    NOTIF_OPENZAAK_SECRET="notif-secret",
-    OPENZAAK_NOTIF_CONFIG_ENABLE=True,
-    OPENZAAK_NOTIF_CLIENT_ID="oz-client-id",
-    OPENZAAK_NOTIF_SECRET="oz-secret",
-    NOTIFICATION_RETRY_CONFIG_ENABLE=True,
-)
 class SetupConfigurationTests(TestCase):
```
Comment (stevenbal) [inline src/nrc/tests/commands/test_setup_configuration.py:23]: Not 100% if we need these full tests tbh

File: docs/installation/configuration/opennotifs_config_cli.rst [hunk 2]
```diff
@@ -21,96 +18,57 @@ You can get the full command documentation with:
...
 
-* ``SITES_CONFIG_ENABLE``: enable Site configuration. Defaults to ``False``.
-* ``OPENNOTIFICATIES_DOMAIN``:  a ``[host]:[port]`` or ``[host]`` value. Required.
-* ``OPENNOTIFICATIES_ORGANIZATION``: name of Open Notificaties organization. Required.
+Services configuration
+----------------------
 
-Authorization configuration
----------------------------
+TODO: add generated documentation for ``zgw_consumers.ServiceConfigurationStep``
```
Comment (SonnyBA) [inline docs/installation/configuration/opennotifs_config_cli.rst:33]: Are these `TODO`'s to be generated later on?

File: docs/installation/configuration/env_config.rst [hunk 3]
```diff
@@ -110,6 +110,7 @@ Optional
 * ``LOG_OUTGOING_REQUESTS_MAX_AGE``: The amount of time after which request logs should be deleted from the database. Defaults to: ``7``.
 * ``SENTRY_DSN``: URL of the sentry project to send error reports to. Default empty, i.e. -> no monitoring set up. Highly recommended to configure this.
 * ``EXTRA_VERIFY_CERTS``: a comma-separated list of paths to certificates to trust, If you're using self-signed certificates for the services that Open Notificaties communicates with, specify the path to those (root) certificates here, rather than disabling SSL certificate verification. Example: ``EXTRA_VERIFY_CERTS=/etc/ssl/root1.crt,/etc/ssl/root2.crt``. Defaults to: ``(empty string)``.
+* ``OPENNOTIFICATIES_DOMAIN``: The domain for this Open Notificaties instance (``[host]:[port]`` or ``[host]``). Defaults to: ``(empty string)``.
```
Comment (SonnyBA) [inline docs/installation/configuration/env_config.rst:113]: Is this change related to this PR? Or was it missing from the file?

File: src/nrc/setup_configuration/site.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+from django_setup_configuration.configuration import BaseConfigurationStep
+
+from .models import SiteConfigModel
+
+
+class SiteConfigurationStep(BaseConfigurationStep[SiteConfigModel]):
+    """
+    Configure the application site/domain.
+
+    **NOTE:** Site configuration will be depreciated
```
Comment (SonnyBA) [inline src/nrc/setup_configuration/site.py:12]: Shouldn't this be upgraded (or disabled)? See [thread](https://github.com/maykinmedia/objects-api/pull/489#discussion_r1877969763)

File: src/nrc/tests/commands/files/setup_config_full.yaml [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+  - identifier: open-zaak
+    secret: G2LIVfXal1J93puQkV3O
+
+# notifications_config_enable: True
+# notifications_config:
+#   # No notifications_api_service necessary, because Open Notificaties doesn't send
+#   # notifications to itself
+#   notification_delivery_max_retries: 5
+#   notification_delivery_retry_backoff: 3
+#   notification_delivery_retry_backoff_max: 30
```
Comment (SonnyBA) [inline src/nrc/tests/commands/files/setup_config_full.yaml:32]:
```suggestion
# notifications_config_enable: True
# notifications_config:
#   # No notifications_api_service necessary, because Open Notificaties doesn't send
#   # notifications to itself
#   notification_delivery_max_retries: 5
#   notification_delivery_retry_backoff: 3
#   notification_delivery_retry_backoff_max: 30
```
This doesn't serve any purpose I guess if it is not possible to configure it

File: src/nrc/tests/commands/test_setup_configuration.py [hunk 6]
```diff
@@ -115,24 +107,14 @@ def test_setup_configuration(self, m):
...
-        command_output = stdout.getvalue()
-
-        self.assertEqual(len(m.request_history), 0)
-        self.assertTrue("Selftest is skipped" in command_output)
+        # with self.subTest("Notifications configured correctly"):
+        #     config = NotificationsConfig.get_solo()
+        #     self.assertEqual(config.notifications_api_service, None)
+        #     self.assertEqual(config.notification_delivery_max_retries, 5)
+        #     self.assertEqual(config.notification_delivery_retry_backoff, 3)
+        #     self.assertEqual(config.notification_delivery_retry_backoff_max, 30)
```
Comment (SonnyBA) [inline src/nrc/tests/commands/test_setup_configuration.py:115]: Similar to the comment [in the test file](https://github.com/open-zaak/open-notificaties/pull/201/files#diff-f6e6534d51058875815a05f879636b3241e9a772af170cb0b0f2043e439ea02d)

File: src/nrc/conf/includes/base.py [hunk 7]
```diff
@@ -132,10 +132,13 @@
...
 #
 SETUP_CONFIGURATION_STEPS = [
-    "nrc.config.site.SiteConfigurationStep",
-    "nrc.config.authorization.AuthorizationStep",
-    "nrc.config.authorization.OpenZaakAuthStep",
-    "nrc.config.notification_retry.NotificationRetryConfigurationStep",
+    "zgw_consumers.contrib.setup_configuration.steps.ServiceConfigurationStep",
+    "vng_api_common.contrib.setup_configuration.steps.JWTSecretsConfigurationStep",
+    "nrc.setup_configuration.authorization.AuthorizationStep",
+    # TODO these need to be implemented later
```
Comment (SonnyBA) [inline src/nrc/conf/includes/base.py:138]: Does this comment contradict the comment[ in the test file](https://github.com/open-zaak/open-notificaties/pull/201/files#diff-f6e6534d51058875815a05f879636b3241e9a772af170cb0b0f2043e439ea02d) about the notification configuration?

File: src/nrc/conf/includes/base.py [hunk 8]
```diff
@@ -132,10 +132,13 @@
...
+    # "notifications_api_common.contrib.setup_configuration.steps.NotificationConfigurationStep",
+    # TODO this should be moved to `django-setup-configuration`
+    # "nrc.config.site.SiteConfigurationStep",
```
Comment (SonnyBA) [inline src/nrc/conf/includes/base.py:141]: This should either be uncomment or removed (with it's configuration step and model)

File: src/nrc/tests/commands/test_setup_configuration.py (hunk 1 above)
Comment (SonnyBA) [reply src/nrc/tests/commands/test_setup_configuration.py:23]: If the steps are not dependent on other steps from within the project itself I don't see any added value of the file

File: docs/installation/configuration/opennotifs_config_cli.rst (hunk 2 above)
Comment (stevenbal) [reply docs/installation/configuration/opennotifs_config_cli.rst:33]: Yeah I wanted to add these placeholders to make sure we don't forget about it, but this will probably happen in january next year

File: docs/installation/configuration/env_config.rst (hunk 3 above)
Comment (stevenbal) [reply docs/installation/configuration/env_config.rst:113]: I thought it was used as a runtime setting as well (https://github.com/open-zaak/open-notificaties/blob/fa08b63c9f269e484daa3651c6ca1ae9c91ec54b/src/nrc/utils/__init__.py#L13), but it seems that the utils in which it's used are part of the existing configuration steps: https://github.com/search?q=repo%3Aopen-zaak%2Fopen-notificaties%20build_absolute_url&type=code

I'll remove it entirely

File: src/nrc/tests/commands/files/setup_config_full.yaml (hunk 5 above)
Comment (stevenbal) [reply src/nrc/tests/commands/files/setup_config_full.yaml:32]: I had these in because I initially wanted to refactor everything at once, but I decided to split it up because this relies on the PR in notifications-api-common.

I'll remove it for now

File: src/nrc/setup_configuration/site.py (hunk 4 above)
Comment (stevenbal) [reply src/nrc/setup_configuration/site.py:12]: I will copy this from your PR in objecttypes once that is merged, to make sure they are the same

Comment (stevenbal) [reply src/nrc/setup_configuration/site.py:12]: I've removed it and will add an issue to add it in another PR

File: src/nrc/conf/includes/base.py (hunk 7 above)
Comment (stevenbal) [reply src/nrc/conf/includes/base.py:138]: This relies on the step from notifications-api-common, which is still a WIP: https://github.com/maykinmedia/notifications-api-common/pull/22

File: src/nrc/conf/includes/base.py (hunk 8 above)
Comment (stevenbal) [reply src/nrc/conf/includes/base.py:141]: I've removed them and added separate issues for them

File: src/nrc/tests/setup_configuration/files/setup_config_sites.yaml
Comment (SonnyBA) [inline src/nrc/tests/setup_configuration/files/setup_config_sites.yaml:1]: This file can be removed, now that the configuration step is also removed

**Review by SonnyBA — APPROVED** [review]

File: src/nrc/setup_configuration/authorization.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+    def execute(self, model: AuthorizationsConfigModel) -> None:
+        auth_config = AuthorizationsConfig.get_solo()
+
+        if auth_config.component != ComponentTypes.nrc:
+            auth_config.component = ComponentTypes.nrc
+
+        service = get_service(model.authorizations_api_service_identifier)
+        auth_config.authorizations_api_service = service
+        auth_config.save(update_fields=("component", "authorizations_api_service"))
```
Comment (annashamray) [inline src/nrc/setup_configuration/authorization.py:45]: This is not enough to use Autorisaties API, you need to also subscribe to changes there, so you need to add missing steps from https://open-notificaties.readthedocs.io/en/stable/installation/configuration/opennotifs_config.html : 

1. configure NotificationsConfig (just like AuthorizationsConfig)
2. subscribe to "autorisaties" channel (use Abonnement or Subscription model for it)

File: src/nrc/setup_configuration/authorization.py [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+def get_service(slug: str) -> Service:
+    """
+    Try to find a Service and re-raise DoesNotExist with the identifier to make debugging
+    easier
+    """
+    try:
+        return Service.objects.get(slug=slug)
+    except Service.DoesNotExist as e:
+        raise Service.DoesNotExist(f"{str(e)} (identifier = {slug})")
```
Comment (annashamray) [inline src/nrc/setup_configuration/authorization.py:18]: Should not you catch this exception in the AuthorizationStep.execute and raise ConfigurationRunFailed ? Or it will be caught later in the `setup_configuration` code? 
Could you please add a test for this case?

**Review by annashamray — DISMISSED** [review]

File: docker/setup_configuration/data.yaml [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+  services:
+  - identifier: autorisaties-api
+    label: Autorisaties API
+    api_root: http://localhost:8000/autorisaties/api/v1/
+    api_type: ac
+    auth_type: zgw
+    client_id: open-notificaties
+    secret: oPMsHCEuoP9Qh8vP06D7
+    user_id: open-notificaties
+    user_representation: Open Notificaties Demodam
```
[the remark this answers is not in the document]
Comment (stevenbal) [reply docker/setup_configuration/data.yaml:12]: figured putting an actual municipality name there would be a bit strange, I'll just remove it

File: src/nrc/setup_configuration/authorization.py [hunk 12]
```diff
@@ -0,0 +1,45 @@
+# SPDX-License-Identifier: EUPL-1.2
+# Copyright (C) 2022 Dimpact
```
[the remark this answers is not in the document]
Comment (stevenbal) [reply src/nrc/setup_configuration/authorization.py:2]: I see I didn't add these for other newly added files, apparently this was never checked in CI? Do we need these comments at all?

File: src/nrc/setup_configuration/authorization.py (hunk 9 above)
Comment (stevenbal) [reply src/nrc/setup_configuration/authorization.py:45]: There is a separate issue for this: https://github.com/open-zaak/open-notificaties/issues/203

File: src/nrc/config/site.py [hunk 13]
```diff
@@ -1,41 +0,0 @@
-from django.conf import settings
```
[the remark this answers is not in the document]
Comment (stevenbal) [reply src/nrc/config/site.py:1]: It's been moved to `django-setup-configuration`, there is a separate PR to reuse it in Open Notificaties: https://github.com/open-zaak/open-notificaties/pull/206

File: src/nrc/config/notification_retry.py [hunk 14]
```diff
@@ -1,37 +0,0 @@
-from django.conf import settings
```
[the remark this answers is not in the document]
Comment (stevenbal) [reply src/nrc/config/notification_retry.py:1]: Also part of https://github.com/open-zaak/open-notificaties/issues/203, this has been added to `notifications-api-common`, so there will be a PR to reuse that step in Open Notificaties

File: src/nrc/setup_configuration/authorization.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (annashamray) [reply src/nrc/setup_configuration/authorization.py:2]: I think we only have a check for it in Open Zaak
So it can be definitely removed here (and I'm not sure why we keep it in OZ, maybe we don't have to)

File: docker/setup_configuration/data.yaml [hunk 15]
```diff
@@ -0,0 +1,36 @@
+zgw_consumers_config_enable: True
+zgw_consumers:
+  services:
+  - identifier: autorisaties-api
+    label: Autorisaties API
+    api_root: http://localhost:8000/autorisaties/api/v1/
```
[the remark this answers is not in the document]
Comment (stevenbal) [reply docker/setup_configuration/data.yaml:6]: I've added Open Zaak to the docker-compose and changed the URLs

File: src/nrc/setup_configuration/authorization.py (hunk 10 above)
Comment (stevenbal) [reply src/nrc/setup_configuration/authorization.py:18]: you're right, I've changed the exception and added a test for it

File: src/nrc/setup_configuration/authorization.py [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+class AuthorizationStep(BaseConfigurationStep[AuthorizationsConfigModel]):
+    """
+    Open Notificaties uses Autorisaties API to check permissions of the clients.
+
+    1. Set up authorization to point to the API
+    2. Add credentials for Open Notifications to request Open Zaak
+
+    Normal mode doesn't change the credentials after its initial creation.
+    If the client_id or secret is changed, run this command with 'overwrite' flag
+    """
```
[the remark this answers is not in the document]
Comment (stevenbal) [reply src/nrc/setup_configuration/authorization.py:30]: I've added some basic documentation to `opennotifs_config_cli.rst`
